Lib/pyclbr.py

```diff
@@ -179,187 +169,113 @@ def _readmodule(module, path, inpackage=None):
-    path is passed to recursive calls of _readmodule.
-    fname becomes o.file.
-    source is tokenized.  Imports cause recursive calls to _readmodule.
-    tree is {} or {'__path__': <submodule search locations>}.
-    inpackage, None or string, is passed to recursive calls of _readmodule.
-
-    The effect of recursive calls is mutation of global _modules.
-    """
-    f = io.StringIO(source)
+class _ClassBrowser(ast.NodeVisitor):
+    def __init__(self, module, path, file, tree, inpackage):
+        self.path = path
+        self.tree = tree
+        self.file = file
+        self.module = module
+        self.inpackage = inpackage
+        self.stack = []
+
+    def visit_ClassDef(self, node):
+        bases = []
+        for base in node.bases:
+            name = ast.unparse(base)
+            if name in self.tree:
+                # We know this super class.
+                bases.append(self.tree[name])
+            elif len(names := name.split(".")) > 1:
+                # Super class form is module.class:
+                # look in module for class.
+                module = names[-2]
+                class_ = names[-1]
+                if module in _modules:
+                    bases.append(_modules[module].get(class_) or name)
+            else:
+                bases.append(name)
+
+        parent = self.stack[-1] if self.stack else None
+        class_ = Class(
+            self.module, node.name, bases, self.file, node.lineno, parent
+        )
+        if parent is None:
+            self.tree[node.name] = class_
+        self.stack.append(class_)
+        self.generic_visit(node)
+        self.stack.pop()
+
+    def visit_Assign(self, node):
+        if (
+            not len(node.targets) == 1
+            or not len(self.stack) > 0
+            or not isinstance(self.stack[-1], Class)
+            or not isinstance(node.targets[0], ast.Name)
+            or not isinstance(node.value, ast.Name)
+            or not isinstance(self.tree.get(node.value.id), Function)
+        ):
+            return
+
+        name = node.targets[0].id
+        child = copy.deepcopy(self.tree[node.value.id])
+        child.parent = self.stack[-1]
+        self.stack[-1]._addchild(name, child)
+        self.stack[-1]._addmethod(name, node.lineno)
+
+    def visit_FunctionDef(self, node, *, is_async=True):
+        parent = self.stack[-1] if self.stack else None
+        function = Function(
+            self.module, node.name, self.file, node.lineno, parent, is_async
+        )
+        if parent is None:
+            self.tree[node.name] = function
+        self.stack.append(function)
+        self.generic_visit(node)
+        self.stack.pop()
+
+    def visit_AsyncFunctionDef(self, node):
+        self.visit_FunctionDef(node, is_async=True)
+
+    def visit_Import(self, node):
+        if node.col_offset != 0:
+            return
+
+        for module in node.names:
+            try:
```

Comment:
I find this code a bit messy, maybe we should wrap or modify `_readmodule` to return a specific exception or `None` if something goes wrong to avoid this double nesting with retry.

Reply:
`_readmodule` is used directly by public `readmodule` and `readmodule_ex` interfaces. I dont think such a change is good because we might be simplify here but we need to add a way to raise that exceptions again in public APIs.

Reply:
> `_readmodule` is used directly by public `readmodule` and `readmodule_ex` interfaces. I dont think such a change is good because we might be simplify here but we need to add a way to raise that exceptions again in public APIs.

Let me think about what would be the best here, I still find this double try a code smell.

Reply:
The double try is messy but is part of the original.    I worry more about having the same result as before.